keyboards/lily58/keymaps/barabas/keymap.c

```diff
@@ -0,0 +1,189 @@
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,  _______, _______,  KC_PLUS, KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS,
+                               _______, _______, _______,  _______, _______,  _______, _______, _______
+),
+
+[_ADJUST] = LAYOUT( \
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,                   XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+    XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX,
+                               _______, _______, _______, _______, _______,  _______, _______, _______
+)
+};
+/* clang-format on */
+
+// Setting ADJUST layer RGB back to default
+void update_tri_layer_RGB(uint8_t layer1, uint8_t layer2, uint8_t layer3) {
+    if (IS_LAYER_ON(layer1) && IS_LAYER_ON(layer2)) {
+        layer_on(layer3);
+    } else {
+        layer_off(layer3);
+    }
+}
+
+// SSD1306 OLED update loop, make sure to enable OLED_DRIVER_ENABLE=yes in rules.mk
+#ifdef OLED_DRIVER_ENABLE
+
+oled_rotation_t oled_init_user(oled_rotation_t rotation) {
+    if (!is_keyboard_master()) return OLED_ROTATION_180;  // flips the display 180 degrees if offhand
+    return rotation;
+}
+
+const char *read_logo(void);
+void        set_keylog(uint16_t keycode, keyrecord_t *record);
+const char *read_keylog(void);
+const char *read_keylogs(void);
+
+#    define RAISE_MASK (1 << _RAISE)
+#    define LOWER_MASK (1 << _LOWER)
+#    define GAME_MASK (1 << _GAME)
+const char *read_layer_state(void) {
+    static char layer_state_str[24];
+
+    switch (layer_state & (~GAME_MASK)) {
+        case (RAISE_MASK | LOWER_MASK):
+            snprintf(layer_state_str, sizeof(layer_state_str), "Layer: Adjust");
+            break;
+        case RAISE_MASK:
+            snprintf(layer_state_str, sizeof(layer_state_str), "Layer: Raise");
+            break;
+        case LOWER_MASK:
+            snprintf(layer_state_str, sizeof(layer_state_str), "Layer: Lower");
+            break;
+        case 0:
+            snprintf(layer_state_str, sizeof(layer_state_str), "Layer: Default");
+            break;
+        default:
+            snprintf(layer_state_str, sizeof(layer_state_str), "Layer: Undef-%ld", layer_state);
+            break;
+    }
+
+    if (layer_state & GAME_MASK) {
+        strncat(layer_state_str, " + Game", sizeof(layer_state_str) - 1);
+    }
+
+    return layer_state_str;
+}
+
+void oled_task_user(void) {
+    if (is_keyboard_master()) {
+        oled_write_ln(read_layer_state(), false);
+        oled_write_ln(read_keylog(), false);
+        oled_write_ln(read_keylogs(), false);
```

Comment:
I know this may be in the default keymap, but converting these to `oled_write_ln_P` instead of the strings and passing them along like this would be a good idea. 

the *_P functions don't use system memory, which has a number of advantages.  Namely, I've personally run into crashes with OLEDs, because of it using all the system memory.  

IIRC, the crkbd has some good examples of how to do this. But basially, don't use snprintf whenever possible.

Reply:
I've changed the `read_layer_state` function to use the _P functions where possible. There is a call to itoa to display the layer state for debugging. I think this is more efficient than printf, but I'm not sure where to look for its implementation.

I think `read_keylog` and `read_keylogs` are out of scope for this PR.